Support for  Angular v21 and a bug fix!

lindolo25 opened the pull request:
Hi @wawyed! This PR implements support for v21 and fixes a recent bug I notice, specific for zoneless apps. I decided to add these changes together to take the least of your time. If instead you will like for me to treat it as 2 separate changes, I'll be happy to split the changes and create merge request individually.

More about the changes:

* Angular v21 support.
* Added downstream test projects for zoneless applications.
* Changes to UIView that introcudes `@if` as replacement for `*ngIf` directive (currently deprecated) and signals specifically for the `_componentRef` property of the class.
* Added test suites that for the specific scenario of the bug. It involved add test environments configured to use zoneless.

**About the bug**: It is only reproducable for applications configure with zoneless change detection or for UIView's instances hosted by a component configured with change detection strategy `onPush`. It involves the projected content of UIViews. here is an example of the components configuration.

```typescript
@Component({
  selector: 'app-root',
  imports: [UIView, AnchorUISref, UISref],
  changeDetection: ChangeDetectionStrategy.OnPush,
  template: `
    <h1>Hello from {{ name }}!</h1>
    <ui-view>
      <a uiSref="root.child">follow this link to open the child state</a>
    </ui-view>
  `,
})
export class AppComponent {
  name = 'UIView Zoneless bug demo';
}

@Component({
  selector: 'app-child',
  imports: [UISref, AnchorUISref],
  changeDetection: ChangeDetectionStrategy.OnPush,
  template: `
    <a uiSref="^">Use this link to go back</a>
    <p>Use the browser's back button to reproduce the bug.</p>
  `,
})
export class AppChildComponent {}

export const routes: Ng2StateDeclaration[] = [
  {
    name: 'root',
    url: '/home',
    component: AppComponent,
  },
  {
    name: 'root.child',
    url: '/child',
    component: AppChildComponent,
  },
];

bootstrapApplication(UIView, {
  providers: [
    provideZonelessChangeDetection(),
    provideUIRouter({
      states: routes,
      otherwise: '/home',
    }),
  ],
});
```
Steps to reproduce: 
1. use the link in the /home page to navigate to the child state. the AppChildComponent will replace the content for UIView.
2. use the link in the /child page to go back to /home. All works fine so far, the projected content for the UIView is restored.
4. use the browser's back button to go back. The projected content is not restored.

Here are a few links to working examples of this bug:
* zoneless: https://stackblitz.com/edit/stackblitz-starters-uonxgsrb
* change detection onPush: https://stackblitz.com/edit/stackblitz-starters-czh5sqlm

Another topic: 

The merge request for `sample-app-angular` that added support for v20 was never merge. I'll be adding a new PR with support for v21. https://github.com/ui-router/sample-app-angular/pull/795

Fixes #1010

lindolo25 committed:
Update Angular packages to v21. This commit also updates minor/mayor
versions for other packages as needed for peer dependency requirements

lindolo25 committed:
Update to compiler options for tests, this fixes failing tests after update to the angular's v21

lindolo25 committed:
Update downstream test porject for typescript to version 5.9

lindolo25 committed:
Replace downstream test project for standalone angular v20 with a v21

lindolo25 committed:
update angular version for downstream project v21-standalone from v20 to
v21. this needs to be done in a separeted step as the cli requires for
the git chnages to be cleared before updating packages.

lindolo25 committed:
Replace downstream test project for v20 to v21.

lindolo25 committed:
Update donwstream_projects file, Angular tests are now v21.

lindolo25 committed:
replace *ngIf with @if for UIView component. it includes transforming
_componentRef to be a readonly signal and updating
ChangeDetectionStrategy to onPush. This fixes issue restoring the
projected content when navigating from a child state to a state with
projected content.

lindolo25 committed:
New test cases to confirm projected content is restored for UIView when
navigating back form a chils state.

lindolo25 committed:
New tests suites for zoneless specific scenarios related to UIView.

lindolo25 committed:
Update to configuration for scripts. added tests fo zoneless and update
the global test to run both zone and zoneless test suites.

lindolo25 committed:
Create zoneless downstream project to test for any regressions with
zoneless aplications.

lindolo25 committed:
reverting the addition of change detection onPush for UIView component.

lindolo25 committed:
Merge remote-tracking branch 'upstream/master' into chore/ng-v21-support

lindolo25 committed:
Bump uirouter/core peer dependency to v6.1.2

lindolo25 commented:
Hi @wawyed and @christopherthielen, I updated this PR with the recent changes to main, it is ready for review. thanks again for your time and Happy new year!

christopherthielen commented:
I'll take a look, thanks for the PR!

Note: I'm not longer using or tracking the Angular project, but I've been updating a bunch of other ui-router packages recently, trying to modernize and consolidate tooling.

christopherthielen commented:
@lindolo25 @wawyed please try this when you can and report any problems

lindolo25 commented:
Hi @christopherthielen, thanks for the help with this PR. 

> I'll take a look, thanks for the PR!
> 
> Note: I'm not longer using or tracking the Angular project, but I've been updating a bunch of other ui-router packages recently, trying to modernize and consolidate tooling.

If you like any help, let me know! I also wanted to ask if it is possible for the documentation to be updated for https://ui-router.github.io. I look at the build process as configured in several of the projects and I think there was an error with the typescript version being incompatible in the Docs container image and the angular project. I can take a look again and create a PR!

christopherthielen commented:
The docgen system seems rather broken at the moment. If you can get it working again, that would be awesome. 

I do have a dream to modernize all the depts of the whole ui router ecosystem and then converge everything into a monorepo. At that point it would be worth looking into different docs paradigm. 

That's a lot though, and may never happen. So if we can resurrect the existing docgen, it would be very useful.

lindolo25 commented:
Hi @christopherthielen, I wanted to update you on some of the troubleshooting I've been doing for the Docs. Indeed the first problem is Typescript compatibility between projects. So far I tested a few changes for projects to make the build work:
* [publish-scripts](https://github.com/ui-router/publish-scripts): Updated the dockerfile to use node v22 and package.json to use typescript v5.9 and typedoc v0.28. 
* [typedoc-plugin-ui-router](https://github.com/christopherthielen/typedoc-plugin-ui-router): updated typedoc to v0.28, and made changes to the implementation of the plugin due to changes in the API.

None of the changes I have are final, that's why I haven't created any pull request. But with them, the docs script of this project builds. Styling is broken and I still have to find out how this build fits with the rest of the projects for the publish. I'll let you know more as I progress!
